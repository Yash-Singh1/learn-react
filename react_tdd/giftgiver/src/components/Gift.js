import React, { Component } from 'react';
import {
  Form,
  FormGroup,
  FormControl,
  FormLabel,
  Button
} from 'react-bootstrap';

class Gift extends Component {
  constructor() {
    super();

    this.state = {
      person: '',
      present: ''
    };
  }

  componentDidMount() {
    this.setState({ person: this.props.person, present: this.props.present });
  }

  render() {
    return (
      <div className="gift">
        <Form>
          <FormGroup>
            <FormLabel>Person</FormLabel>
            <FormControl
              className="input-person"
              value={this.state.person}
              onChange={async (event) => {
                await this.setState({ person: event.target.value });
                this.props.modifyGift(this.props.gift.id, this.state);
              }}
            />
          </FormGroup>
          <FormGroup>
            <FormLabel>Present</FormLabel>
            <FormControl
              className="input-present"
              value={this.state.present}
              onChange={async (event) => {
                await this.setState({ present: event.target.value });
                this.props.modifyGift(this.props.gift.id, this.state);
              }}
            />
          </FormGroup>
        </Form>
        <Button
          className="btn-remove"
          onClick={() => this.props.removeGift(this.props.gift.id)}
        >
          Remove Gift
        </Button>
      </div>
    );
  }
}

export default Gift;
