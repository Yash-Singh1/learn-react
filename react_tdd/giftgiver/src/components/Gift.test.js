import React from 'react';
import { shallow } from 'enzyme';
import Gift from './Gift';

describe('Gift', () => {
  const mockRemove = jest.fn();
  const mockModify = jest.fn();
  const id = 1;
  const props = {
    gift: { id },
    removeGift: mockRemove,
    modifyGift: mockModify,
    person: '',
    present: ''
  };
  const gift = shallow(<Gift {...props} />);

  it('renders properly', () => {
    expect(gift).toMatchSnapshot();
  });

  it('accepts a different person and present prop', () => {
    const differentGift = shallow(<Gift person="John" present="A tree" />);
    differentGift.instance().componentDidMount();
    expect(differentGift.state()).toEqual({
      person: 'John',
      present: 'A tree'
    });
  });

  it('initializes a person and a present in `state`', () => {
    expect(gift.state()).toEqual({ person: '', present: '' });
  });

  describe('when typing in the person input', () => {
    const person = 'Uncle';

    beforeEach(() => {
      gift
        .find('.input-person')
        .simulate('change', { target: { value: person } });
    });

    afterEach(() => {
      gift.find('.input-person').simulate('change', { target: { value: '' } });
    });

    it('updates the person in `state`', () => {
      expect(gift.state().person).toEqual(person);
    });

    it('calls the modify gift callback', () => {
      expect(mockModify).toHaveBeenCalledWith(id, {
        person: 'Uncle',
        present: ''
      });
    });
  });

  describe('when typing in the present input', () => {
    const present = 'Teddy Bear';

    beforeEach(() => {
      gift
        .find('.input-present')
        .simulate('change', { target: { value: present } });
    });

    it('updates the present in `state`', () => {
      expect(gift.state().present).toEqual(present);
    });

    it('calls the modify gift callback', () => {
      expect(mockModify).toHaveBeenCalledWith(id, {
        person: '',
        present: 'Teddy Bear'
      });
    });
  });

  describe('when clicking the remove gift button', () => {
    beforeEach(() => {
      gift.find('.btn-remove').simulate('click');
    });

    it('calls the remove gift callback', () => {
      expect(mockRemove).toHaveBeenCalledWith(id);
    });
  });
});
