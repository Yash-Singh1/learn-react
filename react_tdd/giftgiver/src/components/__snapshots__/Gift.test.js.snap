// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gift renders properly 1`] = `
<div
  className="gift"
>
  <Form
    inline={false}
  >
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Person
      </FormLabel>
      <FormControl
        className="input-person"
        onChange={[Function]}
        value=""
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Present
      </FormLabel>
      <FormControl
        className="input-present"
        onChange={[Function]}
        value=""
      />
    </FormGroup>
  </Form>
  <Button
    active={false}
    className="btn-remove"
    disabled={false}
    onClick={[Function]}
    variant="primary"
  >
    Remove Gift
  </Button>
</div>
`;
